src/kronecker.jl

```diff
@@ -0,0 +1,38 @@
+import Base.kron
+
+export kron, kron!
+
+function kron!(A::CUDA.CuDeviceArray{<:Number},B::CUDA.CuDeviceArray{<:Number},C::CUDA.CuDeviceArray{<:Number})
+
+    C_rows = size(C,1)
+    B_rows = size(B,1)
+    B_cols = size(B,2)
+
+    index1 = (CUDA.blockIdx().x - 1) * CUDA.blockDim().x + CUDA.threadIdx().x
+    stride1 = CUDA.blockDim().x * CUDA.gridDim().x
+
+    index2 = (CUDA.blockIdx().y - 1) * CUDA.blockDim().y + CUDA.threadIdx().y
+    stride2 = CUDA.blockDim().y * CUDA.gridDim().y
+
+    index3 = (CUDA.blockIdx().z - 1) * CUDA.blockDim().z + CUDA.threadIdx().z
+    stride3 = CUDA.blockDim().z * CUDA.gridDim().z
+
+    @inbounds for j = index1:stride1:size(A,2), l = index2:stride2:B_cols, i = index3:stride3:size(A,1)
+        aij = A[i,j]
+        @inbounds for k = 1:B_rows
+            C[(B_rows*(i-1) + k) + C_rows*((B_cols*(j-1) + l)-1)] = aij*B[k,l]
+        end
+    end
+end
+
+function kron(A::CUDA.CuArray{S},B::CUDA.CuArray{T}) where {S,T <: Number}
+
+	col = size(A,2)*size(B,2)
+	if col == one(col)
+		C = CUDA.zeros(promote_type(S,T),size(A,1)*size(B,1))
+	else
+		C = CUDA.zeros(promote_type(S,T),size(A,1)*size(B,1),col)
+	end
+	CUDA.@cuda kron!(A,B,C)
```

Comment:
This launches only a single thread?

Reply:
Apologies - Would making the change below combined with removing CUDA module prefixes be good?  If so, I will update with the necessary tests/changes.

`    nthreads = (8,8,8)
    nelements = size(A,1)*size(B,1)*col
    nblocks = cld(nelements, prod(nthreads))

    @cuda threads=nthreads blocks=nblocks kron!(A,B,C)
    return C
end`

Reply:
Generally it's better to verify the number of threads you can launch against device limits by using the occupancy API or kernel introspection. You want to launch 8*8*8=512 here, which may not work on every device.

```
julia> k = @cuda launch=false kernel()
CUDA.HostKernel{typeof(kernel), Tuple{}}(kernel, CuContext(0x0000000002f43b40, instance 5ace57dbb19a7ff), CuModule(Ptr{Nothing} @0x0000000005abe490, CuContext(0x0000000002f43b40, instance 5ace57dbb19a7ff)), CuFunction(Ptr{Nothing} @0x0000000005e2fe80, CuModule(Ptr{Nothing} @0x0000000005abe490, CuContext(0x0000000002f43b40, instance 5ace57dbb19a7ff))))

julia> CUDA.maxthreads(k)
1024
```

Reply:
Would it be appropriate to default to launching the kernel in kron with threads only in 1 dimension to simplify not needing to calculate 3D blocks?  Or am I overlooking functionality that does that?

Having the kernel as a standalone function still allows the end user to do that if they wish.

Reply:
Yeah, but I don't think users will end up using the inner kernel function, so maybe it's best to make everything 1D then.